### ISSUE TITLE
HW file error

### ISSUE THREAD

**Issue by csecht** [Ricks-Lab/amdgpu-utils#20]
I just set up a new Ubuntu 18.04, Linux kernel 4.15 host with two Rx570, and loaded AMDGPU 18.4 drivers (not 18.5, because I first loaded AMDGPU drivers on a Ubuntu 16.04 system, which I subsequently upgraded to Ubuntu 18.04 - that was the only way I could figure out how to retain the 4.15 kernel, because the 4.18 kernel doesn't work with AMDGPU drivers). 
With `ampgpu-ls` from today's `amdgpu-utils` I see the following errors
```
$ ./amdgpu-ls
AMD Wattman features enabled: 0xffff7fff
amdgpu version: 18.40-673869
2 AMD GPUs detected, 2 may be compatible, checking...
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_max
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_max
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_target
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_target
2 are confirmed compatible.
....etc...
```
And the contents of those two folders
```
$ ls /sys/class/drm/card1/device/hwmon/hwmon3
device      in0_label  power1_average  power1_cap_min  pwm1_max   temp1_crit       uevent
fan1_input  name       power1_cap      pwm1            pwm1_min   temp1_crit_hyst
in0_input   power      power1_cap_max  pwm1_enable     subsystem  temp1_input

$ ls /sys/class/drm/card2/device/hwmon/hwmon4
device      in0_label  power1_average  power1_cap_min  pwm1_max   temp1_crit       uevent
fan1_input  name       power1_cap      pwm1            pwm1_min   temp1_crit_hyst
in0_input   power      power1_cap_max  pwm1_enable     subsystem  temp1_inputt
```

I can, however, control the fan speeds with`amdgpu-pac --execute`, which I verified with `amdgpu-monitor`
```
$ ./amdgpu-pac --execute
(amdgpu-pac:3167): dbind-WARNING **: 15:28:12.776: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
AMD Wattman features enabled: 0xffff7fff
amdgpu version: 18.40-673869
2 AMD GPUs detected, 2 may be compatible, checking...
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_max
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_max
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_target
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_target
2 are confirmed compatible.

Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_target
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_target
Batch file completed: /home/craig/Desktop/amdgpu-utils-master/pac_writer_bf254171022d44a4aa8e4fcdc8ca8094.sh
Writing 2 changes to GPU /sys/class/drm/card1/device/
+ sudo sh -c echo '1' >  /sys/class/drm/card1/device/hwmon/hwmon3/pwm1_enable
[sudo] password for craig: 
+ sudo sh -c echo '102' >  /sys/class/drm/card1/device/hwmon/hwmon3/pwm1
+ sudo sh -c echo 'manual' >  /sys/class/drm/card1/device/power_dpm_force_performance_level
+ sudo sh -c echo '4' >  /sys/class/drm/card1/device/pp_power_profile_mode
PAC execution complete.
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_target
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_target
Batch file completed: /home/craig/Desktop/amdgpu-utils-master/pac_writer_2e23c00831eb4228b2e67df8603f52d5.sh
Writing 2 changes to GPU /sys/class/drm/card2/device/
+ sudo sh -c echo '1' >  /sys/class/drm/card2/device/hwmon/hwmon4/pwm1_enable
+ sudo sh -c echo '102' >  /sys/class/drm/card2/device/hwmon/hwmon4/pwm1
+ sudo sh -c echo 'manual' >  /sys/class/drm/card2/device/power_dpm_force_performance_level
+ sudo sh -c echo '4' >  /sys/class/drm/card2/device/pp_power_profile_mode
PAC execution complete.
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card1/device/hwmon/hwmon3/fan1_target
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_enable
Error: HW file doesn't exist: /sys/class/drm/card2/device/hwmon/hwmon4/fan1_target
```
So, are the HW file errors anything to be concerned with since PAC seems to be working?
Also, my PC build has a PWM hub with two case fans connected to it. Is that related to the errors?
Do I need to upgrade to the 18.5 AMD drivers?
What about that dbind-WARNING?

**Comment by Ricks-Lab** [comment]
@csecht I suspect it is related to the degree of implementation in older driver releases. I found that the latest release implements features not found earlier. Looks like amdgpu is under active development. 

I suspect that these files are not needed for current amdgpu-utils. I will need some time to dig into the code, but I will look to see if theses reads can be disabled if not needed. 

Not sure about the dbind error. I will have to research it.

**Comment by csecht** [comment]
Ah, I just remembered, in this new host I loaded AMDGPU 18.4, not 18.5. Everything up until now was with 18.5 on my other host. I retained 18.4 following a Ubuntu 16.04 upgrade to 18.04, saw that the amdgpu drivers were working, and thought why fix it if it ain’t broke.  But maybe something is broke. I’ll try uninstalling the amdgpu 18.4 package and then install 18.5.

**Comment by Ricks-Lab** [comment]
In checking this out, I found a bug in how the compatible flag is being set. If it is fixed then this error would cause the card to be classified as not compatible, which would be more of a problem. I will fix the compatible flag and for non-critical parameters, I will not set compatible to False, but will instead disable the reading of that sensor. I have several other changes in the works, so it will take some time.

**Comment by Ricks-Lab** [comment]
I have made some modifications on master to address this issue.  My first attempt was to disable if the file was missing.  I tested on my Nano's, where I have a similar issue as yours, and found it was still an issue.  I had to do try/except by sensor so I could disable problem sensors.  Now, you should only see a warning message once per sensor, then it is disabled.  This is only the case for non-critical sensors (like fan sensors, as pwm sensors are used for monitor and control).  For critical sensors, these types of issues will flag the card as incompatible.  Let me know how it works out.

The code on master is not production worthy.  Please only use for testing.  I am working on the plotting function which is kinda working, but needs to be rewritten.  This first try was just my way of figuring things out.

**Comment by csecht** [comment]
Yes, the HW file errors and warnings now appear only at the top of the -ls report. And now there is only the .../fan1_max error msg, while  .../fan1_enable and .../fan1_target are listed as warnings. 
As an aside, I searched but didn't find, if I were to update amdgpu All-Open 18.4 to 18.5 do I need to uninstall 18.4 first or just go ahead with a 18.5 install? And would OpenCL need to be replaced or updated as well?

**Comment by Ricks-Lab** [comment]
@csecht 
I am preparing to release v2.5.0 which includes these fixes and the implementation of a plot feature to give a real-time plot of GPU parameters.  It is the --plot option of amdgpu-monitor.  It does have a limitation of locking up after running for an extended period of time (hours).  I have updated the user guide with basic info on it.  Let me know what you think.

**Comment by csecht** [comment]
I wan't able to get --plot to run because something is broken with my Python.
I tried to install python3-pip, but got errors:
```Unpacking python3-lib2to3 (3.6.7-1~18.04) ...
dpkg: error processing archive /tmp/apt-dpkg-install-pD1ES8/2-python3-lib2to3_3.6.7-1~18.04_all.deb (--unpack):
 trying to overwrite '/usr/lib/python3.6/lib2to3/Grammar.txt', which is also in package libpython3.6-stdlib:amd64 3.6.8-1+xenial1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Preparing to unpack .../3-python3-distutils_3.6.7-1~18.04_all.deb ...
Unpacking python3-distutils (3.6.7-1~18.04) ...
dpkg: error processing archive /tmp/apt-dpkg-install-pD1ES8/3-python3-distutils_3.6.7-1~18.04_all.deb (--unpack):
 trying to overwrite '/usr/lib/python3.6/distutils/README', which is also in package libpython3.6-stdlib:amd64 3.6.8-1+xenial1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
```
Nevertheless, I went ahead and had pip3 install pandas without errors, but, of course, still couldn't bring up the plotter.
So I tried this cure-all:
```
~$ sudo sh -c "apt-get update;apt-get dist-upgrade;apt-get autoremove;apt-get autoclean"
```
which uncovered some (the same) unmet dependencies:
 ```
You might want to run 'apt --fix-broken install' to correct these.
The following packages have unmet dependencies:
 python3-pip : Depends: python3-distutils but it is not installed
               Recommends: python3-dev (>= 3.2) but it is not installed
 python3-setuptools : Depends: python3-distutils but it is not installed
E: Unmet dependencies. Try 'apt --fix-broken install' with no packages (or specify a solution).
...
```
But then,
```
~$ sudo apt --fix-broken install
```
just repeats the initial two errors, leaving me stuck in a circle I don't know how to get out of.

**Comment by Ricks-Lab** [comment]
This has happened to me a few times, probably when having issues with driver installing/uninstalling. Unfortunately, the only solution I worked out was a clean install of the OS. I have gotten very efficient at doing this. All my Linux systems now have two drives, separating os from home to make this easier.

**Comment by Ricks-Lab** [comment]
In both cases I first tried this method to install pandas, but it didn’t work afterward, but maybe it makes a difference:

sudo apt-get install python3-pandas

**Comment by csecht** [comment]
Ugh.  I did a clean install of Ubuntu 18.04 and got python3-pip and pandas installed just fine. But I haven't been able to get amdgpu drivers installed.  I've tried amdgpu 18.5 and 18.4 with both the Linux 4.15 and 4.18 kernels, but no luck. In the past, the only way I've gotten amdgpu drivers installed is when I've upgraded to Ubuntu 18.04 from 16.04 (which retains the Linux 4.15 kernel and apparently some other piece of magic). With that approach either amdgpu 18.4 or 18.5 works.  I'm hesitant to try it again because that approach led to my Python unmet dependency problems (although I suspect those problems were from a failed attempt to upgrade to Python3 under Ubuntu 16.04). There's a discussion at E@H about Ubuntu-amdgpu driver compatibility.  I'll read that over and try again tomorrow.

**Comment by Ricks-Lab** [comment]
Did you do a clean install of 18.04.1 or 18.04.2?  Definitely amdgpu install has issues with 18.04.2. I have had no problems with 18.04.1, then do a dist-upgrade.

**Comment by csecht** [comment]
Hmm, yes, thanks, the Ubuntu install that couldn't load amdgu was 18.04.2.  I'll try a clean install with the 18.04.1 version next. 
On my other host that has amdgpu 18.5 drivers installed, the one that was upgraded from Ubuntu 16.04, it is running 18.04.2, but with the Linux 4.15.0-43-generic kernel. However, even though it also has python3-pip and pandas installed (for pandas, I had to use your suggestion of sudo apt-get install python3-pandas), I cannot get *amdgpu-monitor --plot* to run; it just loads the standard monitor.
I just realized that the kernel on the host that can't load amdgpu is Linux 4.15.0-41500-generic, so it's not the same version as on the currently working host.  Tricky business.

**Comment by Ricks-Lab** [comment]
In order to run plot you must use both the —gui option and —plot option. You should see some messages in the window where you executed the command on the status or errors in starting plot.  From my experience, the apt install of python3-pandas is not sufficient for plot to run.  It takes pip install.

**Comment by csecht** [comment]
Dang, sorry, didn't use --gui option.  Still no luck though, because I can't install pip3 pandas.  I get these exceptions and error:
```
$ pip3 install pandas
Collecting pandas
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdcc065358>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/pandas/
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdcc0654e0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/pandas/
  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdcc065438>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/pandas/
  Using cached https://files.pythonhosted.org/packages/19/74/e50234bc82c553fecdbd566d8650801e3fe2d6d8c8d940638e3d8a7c5522/pandas-0.24.2-cp36-cp36m-manylinux1_x86_64.whl
Collecting numpy>=1.12.0 (from pandas)
  Using cached https://files.pythonhosted.org/packages/35/d5/4f8410ac303e690144f0a0603c4b8fd3b986feb2749c435f7cdbb288f17e/numpy-1.16.2-cp36-cp36m-manylinux1_x86_64.whl
Collecting pytz>=2011k (from pandas)
  Using cached https://files.pythonhosted.org/packages/3d/73/fe30c2daaaa0713420d0382b16fbb761409f532c56bdcc514bf7b6262bb6/pytz-2019.1-py2.py3-none-any.whl
Collecting python-dateutil>=2.5.0 (from pandas)
  Using cached https://files.pythonhosted.org/packages/41/17/c62faccbfbd163c7f57f3844689e3a78bae1f403648a6afb1d0866d87fbb/python_dateutil-2.8.0-py2.py3-none-any.whl
Collecting six>=1.5 (from python-dateutil>=2.5.0->pandas)
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdc9f91dd8>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/six/
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdc9ef1a58>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fbdc9ef1a90>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
  Downloading https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Installing collected packages: numpy, pytz, six, python-dateutil, pandas
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 360, in run
    prefix=options.prefix_path,
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 851, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 1064, in move_wheel_files
    isolated=self.isolated,
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 247, in move_wheel_files
    prefix=prefix,
  File "/usr/lib/python3/dist-packages/pip/locations.py", line 153, in distutils_scheme
    i.finalize_options()
  File "/usr/lib/python3.6/distutils/command/install.py", line 351, in finalize_options
    self.create_home_path()
  File "/usr/lib/python3.6/distutils/command/install.py", line 581, in create_home_path
    os.makedirs(path, 0o700)
  File "/usr/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/usr/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/home/craig/.local/lib/python3.6'
```

**Comment by Ricks-Lab** [comment]
I suggest seeing if you can get it to work with a clean install of 18.04.1, then worry about the other system that was upgraded from 16.04.  The amdgpu driver packages seem to run about a quarter behind the Ubuntu releases, so I suspect the next driver release will install on 18.04.2.

I have made some updates to the way amdgpu-plot works.  It can now be run on its own in addition to being called by amdgpu-monitor with the --plot option.  When using the amdgpu-monitor with --plot specified, --gui is assumed even if not specified.  I suggest not running both plot and monitor independently, since it will result in twice as many driver reads.

**Comment by csecht** [comment]
Yes, that worked. Thanks much for the 18.04.01 tip. On this new host,  I installed Ubuntu instead of Lubuntu and like it better. Drivers, Python2-pip, pandas installed without error, but now can't run plot because of this:
```
$ ./amdgpu-monitor --plot
AMD Wattman features enabled: 0xffff7fff
amdgpu version: 18.50-756341
2 AMD GPUs detected, 2 may be compatible, checking...
2 are confirmed compatible.

Traceback (most recent call last):
  File "/home/craig/Desktop/amdgpu-utils-master/amdgpu-plot", line 31, in <module>
    from matplotlib.ticker import AutoLocator
ModuleNotFoundError: No module named 'matplotlib'
amdgpu-plot has closed
```
My old host with the cobbled together system I'll deal with separately.

**Comment by csecht** [comment]
*I meant ....Python3-pip...

**Comment by Ricks-Lab** [comment]
I thought matplotlib was installed by default. Maybe it needs to be installed also.  I will need to google it...

**Comment by Ricks-Lab** [comment]
I think this worked.  I didn’t have to work out pip3 install until pandas. 
sudo apt-get install python3-matplotlib

**Comment by Ricks-Lab** [comment]
I also modified amdgpu-plot to give an indication of how to install matplotlib if missing.

**Comment by csecht** [comment]
Okay, I got matplotlib loaded.  Maybe the issue was that I had done a minimal install of Ubuntu?  Now the plot window opens, but nothing gets plotted.  The terminal has this:
```
amdgpu-plot waiting for initial data................
matplotlib error, stack size is 3
Stopping reader thread
exit stack size:  4
```
The GUI monitor ticks away its readouts as normal.

**Comment by Ricks-Lab** [comment]
Both of my systems have 4 and 5 of the same GPU each, which I think is different from your multiple type GPU system. Can you run monitor with the logging option so I can test your situation on my system? I should be able to pipe your log into plot here to simuale the way monitor works with plot.

**Comment by Ricks-Lab** [comment]
I created a log file with '-1' temperature values but could not duplicate the issue.  But I still put in a fix to convert unset values to 'NA' which should be appropriately dealt with by numpy.  I also add *--simlog* option to plot so a log file can be piped into it for testing.

**Comment by Ricks-Lab** [comment]
I just found an error in my original simlog implementation, so my testing with the bad data wasn't using the logfile, so those tests were invalid.  I suspect that the problem is now solved, but it would be difficult to go back to pre changes to verify.  Let me know if the latest works.  If not, please provide a monitor log.

**Comment by csecht** [comment]
Nope, no change. Attached is a few minutes of monitor logging.  After about a minute I used PAC to change the fan PWM of Card2 to default so that its speed changes and its temperature increases (but only by 1 C).
[log_monitor_0421_122149.txt](https://github.com/Ricks-Lab/amdgpu-utils/files/3101189/log_monitor_0421_122149.txt)

**Comment by Ricks-Lab** [comment]
Thanks for providing the log.  I was able to identify the problem and have uploaded a updated version.

**Comment by csecht** [comment]
Wheeee! That's so cool.  I like the tags on the end of plot lines showing current values. Nice.

**Comment by Ricks-Lab** [comment]
Looks like AMD just release new drivers for 18.04.2.  Have you tried them?

**Comment by Ricks-Lab** [comment]
@csecht 
I have tweaked the y-axis max values to reduce the overlap of plot lines.  Also, fixed optimization for 3 GPU system.  I have update the readme and userguide with quick descriptions.  I plan to release this weekend, so let me know of any issues and feel free to update.  Thanks!

**Comment by csecht** [comment]
Not yet.

**Comment by csecht** [comment]
I found that *amdgpu-plot --stdin* stalls in the terminal at 
```
amdgpu-plot is waiting for initial data...
```
 with dots that keep on coming with not plot window launching.  The --stdin option is mentioned in the README.md file, but not in the USER_GUIDE.  I'm not sure whether I'm using it right.

I find that plot lines can overlap on the initial display thus hiding a line (usually the temp plots), but that it can be easily remedied by unselecting the parameter line that's blocking it.  It took me a while to figure that out though. I like the feature of being able to toggle display of plot parameters.

**Comment by Ricks-Lab** [comment]
The *—stdin* is used by amdgpu-monitor where monitor will write to plot’s stdin. It can also be used in simulation mode, like I did with the log file you sent me:
‘’’
cat logfile | ./amdgpu-plot —simlog —stdin —sleep 2
‘’’

**Comment by Ricks-Lab** [comment]
> I find that plot lines can overlap on the initial display thus hiding a line (usually the temp plots), but that it can be easily remedied by unselecting the parameter line that's blocking it. It took me a while to figure that out though. I like the feature of being able to toggle display of plot parameters.

With 6 plots in a small graphing area, it is difficult to not have overlapping lines.  The changed in my last update we to make sure the right axis highest point was lower than max of that axis, otherwise, the plot would automatically have the highest value plots of both axes overlap.  I had also made a similar change the right axis lower limit, but not as large as the higher limit.

**Comment by Ricks-Lab** [comment]
Improvements and fixes discussed here are released in version 2.5.0.
